### ISSUE TITLE
Automatically Submitting after Re-Initialize [6.4.3]

### ISSUE THREAD

**Issue by r-f-j-thomas** [erikras/redux-form#2449]
I have a form which contains search parameter fields, and have been attempting to hook this up to react-router so that the search terms are kept in the URL hash, and so that changing the URL triggers a query with the updated search params.

My approach has been to tie the URL querystring to the form's initialValues (`initialValues={this.props.location.query}`) and setting `enableReinitialize`. Altering the URL therefore causes the form to re-initialize with the new query - this works as expected.

However, when this happens I want the form to automatically submit. I have tried to do this using componentDidUpdate:
```
componentDidUpdate(prevProps) {
	if (!this.props.submitting && !areEqual(prevProps.initialValues, this.props.initialValues)) {
		this.props.submit();
	}
}
```

However this is causing the form to submit with the *previous* values. I can see that `@@redux-form/INITIALIZE` was fired before `@@redux-form/SUBMIT`, so it looks as though it should be using the new values.

After the submit() is fired I can see that the submitIfNeeded function in reduxForm.js has both this.props.initialValues and nextProps.values set to the new values, but this.props.values (which is passed to onSubmit) is still set to the old values. It appears that redux-form initializes but doesn't update props.values until the next render.

It seems strange to me that firing submit after firing initialize (I have tried explicitly firing initialize rather than using enableReinitialize) would not use the new initialValues, but I'm not sure I know enough to suggest there is a bug in the library. Is something off about what Ive done here? Is there a better way to do it?

**Comment by danielrob** [comment]
I'm wondering if this is related to @SBRK's investigation over at #3286?

**Comment by SBRK** [comment]
Yes, sounds like the kind of problem I had. Try calling submit in a `setTimeout`, and see if it works

**Comment by r-f-j-thomas** [comment]
Yes, if I recall setTimeout was one of the fixes I ended up with. Another option in this case was to dispatch the startSubmit/stopSubmit actions manually and call onSubmit directly. Neither is particularly pleasant.

**Comment by paraboom** [comment]
Hey, guys, I use `setTimeout` also to submit the form right after initialization, but that way is a little bit hacky, isn't it? I think there should be the right way.
